File: codes/swift/005_objects.swift
import Foundation

typealias MyObject = [String: Any?]
typealias MyArray = [Any?]

func jsonStringify(_ anything: Any? = nil, pretty: Bool = false, indent: String = "    ") -> String {
    var indentLevel = 0
    func jsonStringifyInner(_ anythingInner: Any?, _ indentInner: String) -> String {
        guard let anythingInner = anythingInner else {
            return "null"
        }
        if let anythingInner = anythingInner as? String {
            return "\"\(anythingInner)\""
        }
        if let anythingInner = anythingInner as? Bool {
            return "\(anythingInner)"
        }
        if let anythingInner = anythingInner as? NSNumber {
            return "\(anythingInner)"
        }
        if let anythingInner = anythingInner as? MyArray {
            if (anythingInner.count == 0) {
                return "[]"
            }
            indentLevel += 1
            var result = ((pretty == true) ? "[\n\(String(repeating: indentInner, count: indentLevel))" : "[")
            for (arrayItemIndex, arrayItem) in anythingInner.enumerated() {
                result += jsonStringifyInner(arrayItem, indentInner)
                if ((arrayItemIndex + 1) != anythingInner.count) {
                    result += ((pretty == true) ? ",\n\(String(repeating: indentInner, count: indentLevel))" : ", ")
                }
            }
            indentLevel -= 1
            result += ((pretty == true) ? "\n\(String(repeating: indentInner, count: indentLevel))]" : "]")
            return result
        }
        if let anythingInner = anythingInner as? MyObject {
            if (anythingInner.count == 0) {
                return "{}"
            }
            indentLevel += 1
            var result = ((pretty == true) ? "{\n\(String(repeating: indentInner, count: indentLevel))" : "{")
            for (objectEntryIndex, (objectKey, objectValue)) in anythingInner.enumerated() {
                result += "\"\(objectKey)\": \(jsonStringifyInner(objectValue, indentInner))"
                if ((objectEntryIndex + 1) != anythingInner.count) {
                    result += ((pretty == true) ? ",\n\(String(repeating: indentInner, count: indentLevel))" : ", ")
                }
            }
            indentLevel -= 1
            result += ((pretty == true) ? "\n\(String(repeating: indentInner, count: indentLevel))}" : "}")
            return result
        }
        return "null"
    }
    return jsonStringifyInner(anything, indent)
}

// in Swift, JavaScript-like Object is called Dictionary

var friend: MyObject = [
    "name": "Alisa",
    "country": "Finland",
    "age": 25
]
print("friend: \(jsonStringify(friend, pretty: true))")

print("friend, get total object keys: \(friend.count)")
// friend, get total object keys: 3

print("friend, get country: \(friend["country"] as? String ?? "null")")
// friend, get country: Finland

print("friend, get country:", { () -> String in
    // this is called optional binding
    if let result = friend["country"] as? String {
        return result
    }
    return "null"
}())
// friend, get country: Finland

print("friend, get country:", { () -> String in
    // this is called optional binding
    guard let result = friend["country"] as? String else {
        return "null"
    }
    return result
}())
// friend, get country: Finland

// iterate over and get each key-value pair
for (objectKey, objectValue) in friend {
    print("friend, for loop, key: \(objectKey), value: \(objectValue ?? "null")")
}
// friend, for loop, key: name, value: Alisa
// friend, for loop, key: country, value: Finland
// friend, for loop, key: age, value: 25

// iterate over and get each key-value pair and object iteration/entry index
for (objectEntryIndex, (objectKey, objectValue)) in friend.enumerated() {
    print("friend, for loop, object iteration/entry index: \(objectEntryIndex), key: \(objectKey), value: \(objectValue ?? "null")")
}
// friend, for loop, object iteration/entry index: 0, key: name, value: Alisa
// friend, for loop, object iteration/entry index: 1, key: country, value: Finland
// friend, for loop, object iteration/entry index: 2, key: age, value: 25

friend["age"] = 27
print("friend: \(jsonStringify(friend, pretty: true))")

friend["gender"] = "Female"
print("friend: \(jsonStringify(friend, pretty: true))")

friend.removeValue(forKey: "country")
print("friend: \(jsonStringify(friend, pretty: true))")

// Computed property names: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#computed_property_names
let deliveryResponseKeyMessage = "message"
var deliveryResponse: MyObject = [
    deliveryResponseKeyMessage: "ok"
]
print("deliveryResponse: \(jsonStringify(deliveryResponse, pretty: true))")
let deliveryResponseKeyStatus = "status"
deliveryResponse[deliveryResponseKeyStatus] = 200
print("deliveryResponse: \(jsonStringify(deliveryResponse, pretty: true))")

File: codes/scala/005_objects.scala
import scala.collection.mutable

def createNewObject(objectEntry: (String, Any)*): Any = (if (objectEntry.isEmpty == true) mutable.Map.empty[String, Any] else mutable.Map(objectEntry: _*))

def isLikeJsArray(anything: Any): Any = {
    if (anything.isInstanceOf[mutable.ArrayBuffer[_]] == true) return true
    return false
}

def getType(anything: Any): Any = {
    // if (isLikeJsArray(anything) == true) return "Array"
    return anything.getClass.getName
}

val isNumeric = ((anything: Any) => (if (anything.isInstanceOf[Byte] || anything.isInstanceOf[Int] || anything.isInstanceOf[Long] || anything.isInstanceOf[Short] || anything.isInstanceOf[Double] || anything.isInstanceOf[Float] || anything.isInstanceOf[BigInt] || anything.isInstanceOf[BigDecimal]) true else false): Boolean)

def jsonStringify(anything: Any = null, pretty: Boolean = false, indent: String = "    "): String = {
    var indentLevel: Int = 0
    def jsonStringifyInner(anythingInner: Any, indentInner: String): String = {
        if (anythingInner == null) return "null"
        if (anythingInner.isInstanceOf[String]) return s"\"${anythingInner}\""
        if (isNumeric(anythingInner) || anythingInner.isInstanceOf[Boolean]) return s"${anythingInner}"
        if (anythingInner.isInstanceOf[mutable.ArrayBuffer[_]]) {
            if (anythingInner.asInstanceOf[mutable.ArrayBuffer[Any]].length == 0) return "[]"
            indentLevel += 1
            var result: String = (if (pretty == true) s"[\n${indentInner * indentLevel}" else "[")
            for ((arrayItem, arrayItemIndex) <- anythingInner.asInstanceOf[mutable.ArrayBuffer[Any]].zipWithIndex) {
                result += jsonStringifyInner(arrayItem, indentInner)
                if ((arrayItemIndex + 1) != anythingInner.asInstanceOf[mutable.ArrayBuffer[Any]].length) result += (if (pretty == true) s",\n${indentInner * indentLevel}" else ", ")
            }
            indentLevel -= 1
            result += (if (pretty == true) s"\n${indentInner * indentLevel}]" else "]")
            return result
        }
        if (anythingInner.isInstanceOf[mutable.Map[_, _]]) {
            if (anythingInner.asInstanceOf[mutable.Map[String, Any]].size == 0) return "{}"
            indentLevel += 1
            var result: String = (if (pretty == true) s"{\n${indentInner * indentLevel}" else "{")
            for (((objectKey, objectValue), objectEntryIndex) <- anythingInner.asInstanceOf[mutable.Map[String, Any]].toArray[Any].zipWithIndex) {
                result += s"\"${objectKey}\": ${jsonStringifyInner(objectValue, indentInner)}"
                if ((objectEntryIndex + 1) != anythingInner.asInstanceOf[mutable.Map[String, Any]].size) result += (if (pretty == true) s",\n${indentInner * indentLevel}" else ", ")
            }
            indentLevel -= 1
            result += (if (pretty == true) s"\n${indentInner * indentLevel}}" else "}")
            return result
        }
        return "null"
    }
    return jsonStringifyInner(anything, indent)
}

// in Scala, JavaScript-like Object is called mutable Map

val friend: Any = mutable.Map[String, Any](
    "name" -> "Alisa",
    "country" -> "Finland",
    "age" -> 25
)
println(s"friend: ${jsonStringify(friend, pretty = true)}")

println(s"friend, get total object keys: ${friend.asInstanceOf[mutable.Map[String, Any]].size}")
// friend, get total object keys: 3

println(s"friend, get country: ${friend.asInstanceOf[mutable.Map[String, Any]]("country")}")
// friend, get country: Finland

// iterate over and get each key-value pair
friend.asInstanceOf[mutable.Map[String, Any]].foreach((objectEntry: Tuple2[String, Any]) => { println(s"friend, forEach loop, key: ${objectEntry._1}, value: ${objectEntry._2}") })
// fruits, forEach loop, key: name, value: Alisa
// fruits, forEach loop, key: country, value: Finland
// fruits, forEach loop, key: age, value: 25

// iterate over and get each key-value pair
friend.asInstanceOf[mutable.Map[String, Any]].foreach((objectEntry: Tuple2[String, Any]) => {
    val (objectKey, objectValue): (String, Any) = (objectEntry._1, objectEntry._2)
    println(s"friend, forEach loop, key: ${objectKey}, value: ${objectValue}")
})
// fruits, forEach loop, key: name, value: Alisa
// fruits, forEach loop, key: country, value: Finland
// fruits, forEach loop, key: age, value: 25

// iterate over and get each key-value pair
friend.asInstanceOf[mutable.Map[String, Any]].foreach { case (objectKey, objectValue) => println(s"friend, forEach loop, key: ${objectKey}, value: ${objectValue}") }
// fruits, forEach loop, key: name, value: Alisa
// fruits, forEach loop, key: country, value: Finland
// fruits, forEach loop, key: age, value: 25

// iterate over and get each key-value pair
friend.asInstanceOf[mutable.Map[String, Any]].foreach(new Function1[Tuple2[String, Any], Unit] {
    def apply(objectEntry: Tuple2[String, Any]): Unit = {
        val (objectKey, objectValue): (String, Any) = (objectEntry._1, objectEntry._2)
        println(s"friend, forEach loop, key: ${objectKey}, value: ${objectValue}")
    }
})
// fruits, forEach loop, key: name, value: Alisa
// fruits, forEach loop, key: country, value: Finland
// fruits, forEach loop, key: age, value: 25

// iterate over and get each key-value pair
for ((objectKey, objectValue) <- friend.asInstanceOf[mutable.Map[String, Any]]) {
    println(s"friend, forEach loop, key: ${objectKey}, value: ${objectValue}")
}
// fruits, forEach loop, key: name, value: Alisa
// fruits, forEach loop, key: country, value: Finland
// fruits, forEach loop, key: age, value: 25

// iterate over and get each key-value pair and object iteration/entry index (the best way)
for (((objectKey, objectValue), objectEntryIndex) <- friend.asInstanceOf[mutable.Map[String, Any]].toArray[Any].zipWithIndex) {
    println(s"friend, forEach loop, object iteration/entry index: ${objectEntryIndex}, key: ${objectKey}, value: ${objectValue}")
}
// friend, forEach loop, object iteration/entry index: 0, key: name, value: Alisa
// friend, forEach loop, object iteration/entry index: 1, key: country, value: Finland
// friend, forEach loop, object iteration/entry index: 2, key: age, value: 25

friend.asInstanceOf[mutable.Map[String, Any]]("age") = 27
println(s"friend: ${jsonStringify(friend, pretty = true)}")

friend.asInstanceOf[mutable.Map[String, Any]]("gender") = "Female"
println(s"friend: ${jsonStringify(friend, pretty = true)}")

friend.asInstanceOf[mutable.Map[String, Any]].update("job", "Streamer")
println(s"friend: ${jsonStringify(friend, pretty = true)}")

friend.asInstanceOf[mutable.Map[String, Any]] += ("address" -> "123 Main Street, Anytown, Finland")
println(s"friend: ${jsonStringify(friend, pretty = true)}")

friend.asInstanceOf[mutable.Map[String, Any]] -= "name"
println(s"friend: ${jsonStringify(friend, pretty = true)}")

friend.asInstanceOf[mutable.Map[String, Any]].remove("country")
println(s"friend: ${jsonStringify(friend, pretty = true)}")

// Computed property names: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#computed_property_names
val deliveryResponseKeyMessage: Any = "message"
val deliveryResponse: Any = mutable.Map[String, Any](
    deliveryResponseKeyMessage.asInstanceOf[String] -> "ok".asInstanceOf[Any]
)
println(s"deliveryResponse: ${jsonStringify(deliveryResponse, pretty = true)}")
val deliveryResponseKeyStatus: Any = "status"
deliveryResponse.asInstanceOf[mutable.Map[String, Any]](deliveryResponseKeyStatus.asInstanceOf[String]) = 200
println(s"deliveryResponse: ${jsonStringify(deliveryResponse, pretty = true)}")
